src/lib/arch/riscv/Instruction_execute.cc

```diff
@@ -827,6 +841,642 @@ void Instruction::execute() {
+      // addresses
+    case Opcode::RISCV_CSRRW: {  // CSRRW rd,csr,rs1
+      // Do nothing to allow progression
+
+      if (metadata.operands[1].reg == 2) {
+        // TODO this won't work properly in OoO core as rounding mode could be
+        // updated before instructions earlier in program order execute causing
+        // them to be rounded incorrectly. Rounding mode needs to be set on
+        // commit and all subsequent instructions must be performed with updated
+        // mode. e.g. this should be performed atomically
+
+        switch (operands[0].get<uint64_t>()) {
+          case 0:
+            fesetround(FE_TONEAREST);
+            break;
+          case 1:
+            fesetround(FE_TOWARDZERO);
+            break;
+          case 2:
+            fesetround(FE_DOWNWARD);
+            break;
+          case 3:
+            fesetround(FE_UPWARD);
+            break;
+          case 4:
+            fesetround(FE_TONEAREST);
+            break;
+          default:
+            // Invalid Case
+            // TODO any subsequent attempt to execute a floating-point operation
+            // with a dynamic rounding mode will raise an illegal instruction
+            // exception.
+            std::cerr << "[SimEng:RISC-V:Execute] Invalid rounding mode"
+                      << std::endl;
+            break;
+        }
+      } else {
+        //        std::cerr << "[SimEng:RISC-V:Execute] Unknown system register:
+        //        "
+        //                  << metadata.operands[1].reg << std::endl;
+      }
+      results[0] = RegisterValue(0, 8);
+      break;
+    }
+    case Opcode::RISCV_CSRRWI: {  // CSRRWI rd,csr,imm
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRS: {  // CSRRS rd,csr,rs1
+      // dummy implementation needs capstone update to recognise system
+      // registers
+
+      results[0] = RegisterValue(static_cast<uint64_t>(0), 8);
+      break;
+    }
+    case Opcode::RISCV_CSRRSI: {  // CSRRSI rd,csr,imm
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRC: {  // CSRRC rd,csr,rs1
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRCI: {  // CSRRCI rd,csr,imm
+      executionNYI();
+      break;
+    }
+
+      // TODO "Apart from transfer operations ... all other floating-point
+      // operations on narrower n-bit operations, n < FLEN, check if the input
+      // operands are correctly NaN-boxed, i.e., all upper FLEN−n bits are 1. If
+      // so, the n least-significant bits of the input are used as the input
+      // value, otherwise the input value is treated as an n-bit canonical NaN."
+
+      // TODO need to take rounding mode into account as these can be set per
+      // instruction https://en.cppreference.com/w/cpp/numeric/fenv/FE_round
+      //      std::fesetround(FE_TONEAREST);
+
+      // Single-Precision Floating-Point (F)
+      // Double-Precision Floating-Point (D)
```

Comment:
Can we take rounding mode into account now? Or is there something blocking this making it a TODO?

Reply:
Rounding modes per instruction would require a potentially slow and slightly messy implementation as this would need to be checked and adjusted for every instruction (and then reset after its execution). Rounding mode info would need to be extracted from the raw bytes, far from impossible but less than ideal. The codes we run at the moment don't seem to be reliant on precise rounding  and so this was left out to reduce complexity in this initial implementation. The TODO remains to keep track of this however

Reply:
Are there a fixed number of rounding modes? And are they always encoded in the same bits for every instruction? 

If so then a simple general function could be implemented which does a bitwise mask on the instruction bytes and returns the rounding mode based on this? If it can be this simple then it would be interesting to see what the performance hit is

Reply:
Implementation done